pym/bob/state.py

```diff
@@ -255,6 +256,49 @@ def setBuildState(self, digest2Dir):
     def getBuildState(self):
         return copy.deepcopy(self.__buildState)
 
+class _BobLock():
+    instance = None
+    def __init__(self):
+        self.__lockfiles = {}
+
+    def getNamedLock(self, name):
+        try:
+            fd = open(name, 'w')
+            fcntl.flock(fd, fcntl.LOCK_EX)
+            self.__lockfiles[name] = fd
+        except OSError as e:
+            from .tty import colorize
+            from sys import stderr
+            print(colorize("Warning: cannot lock " + name + ": "+str(e), "33"),
+                 file=stderr)
+
+    def releaseNamedLock(self, name):
+        if self.__lockfiles[name]:
+            try:
+                fcntl.flock(self.__lockfiles[name], fcntl.LOCK_UN)
+                self.__lockfiles[name].close()
+            except FileNotFoundError:
+                from .tty import colorize
+                from sys import stderr
+                print(colorize("Warning: lock file " + name + " was deleted while Bob was still running!", "33"),
+                    file=stderr)
+            except OSError as e:
+                from .tty import colorize
+                from sys import stderr
+                print(colorize("Warning: cannot unlock workspace: "+str(e), "33"),
+                    file=stderr)
+            finally:
+                del self.__lockfiles[name]
+
+    def finalize(self):
+        for name, fd in self.__lockfiles.items():
```

Comment:
This will fail. While iterating the object must not change. A solution could be to store the keys in a list and iterate that:

```python
for name in list(self.__lockfiles.keys()):
    ...
```